File: 0011-container-with-most-water/0011-container-with-most-water.js
/**
 * @param {number[]} height
 * @return {number}
 */
var maxArea = function(height) {
    let lPt = 0
    let rPt = height.length-1
    let maxArea = 0

    while(lPt < rPt){
        const w = rPt-lPt
        const h = Math.min(height[lPt], height[rPt])
        maxArea = Math.max(maxArea, w*h)   
        if(height[lPt] < height[rPt]){
            lPt++
        } else {
            rPt--
        }  
    }
    
    return maxArea
};

// left pointer, right pointer
// max area 0
// while left pointer is less than right pointer
// if left pointer height is less than right pointer, left ++
// max area is max between maxArea and right pointer - left pointer * min of heights
// if right pointer height is less than left pointer, right --
// update max area

File: 0011-container-with-most-water/0011-container-with-most-water.py
class Solution:
    def maxArea(self, height: List[int]) -> int:
        maxArea = 0
        l = 0
        r = len(height)-1

        while l < r:
            w = r-l
            h = min(height[l], height[r])
            a = w * h
            maxArea = max(maxArea, a)
            if height[l] < height[r]:
                l += 1
            elif height[l] > height[r]:
                r -= 1
            else:
                l += 1
                r -= 1
        return maxArea
        